heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,547 @@
+          .iterator();
+      while (it.hasNext()) {
+        Map.Entry<Protos.Offer, TaskResources> kv = it.next();
+        Protos.Offer offer = kv.getKey();
+        TaskResources resources = kv.getValue();
+
+        if (resources.canSatisfy(neededResources)) {
+          resources.consume(neededResources);
+
+          // Construct the list of free ports to use for a heron executor.
+          List<Integer> freePorts = new ArrayList<>();
+          for (int port = (int) (neededResources.getPortsHold().get(0).rangeStart);
+               port <= (int) (neededResources.getPortsHold().get(0).rangeEnd); port++) {
+            freePorts.add(port);
+          }
+
+          // Add the tasks
+          tasks.add(new LaunchableTask(taskId, baseContainer, offer, freePorts));
+
+          // Set the flag
+          isMatched = true;
+
+          // Matched baseContainer, break;
+          break;
+        }
+      }
+
+      if (!isMatched) {
+        LOG.info(String.format("Insufficient resources remaining for baseContainer: %s, "
+                + "will append to queue. Need: [%s]",
+            taskId, neededResources.toString()));
+        toScheduleTasks.add(taskId);
+
+        // Offer will not be able to satisfy resources needed for any rest container,
+        // since all containers are homogeneous.
+        // Break the loop
+        break;
+      }
+    }
+
+    return tasks;
+  }
+
+  /**
+   * Launch tasks
+   *
+   * @param tasks list of LaunchableTask
+   */
+  protected void launchTasks(List<LaunchableTask> tasks) {
+    // Group the tasks by offer
+    Map<Protos.Offer, List<LaunchableTask>> tasksGroupByOffer = new HashMap<>();
+    for (LaunchableTask task : tasks) {
+      List<LaunchableTask> subTasks = tasksGroupByOffer.get(task.offer);
+      if (subTasks == null) {
+        subTasks = new LinkedList<>();
+      }
+      subTasks.add(task);
+      tasksGroupByOffer.put(task.offer, subTasks);
+    }
+
+    // Construct the Mesos TaskInfo to launch
+    for (Map.Entry<Protos.Offer, List<LaunchableTask>> kv : tasksGroupByOffer.entrySet()) {
+      // We would launch taks for the same offer at one shot
+      Protos.Offer offer = kv.getKey();
+      List<LaunchableTask> subTasks = kv.getValue();
+
+      List<Protos.TaskInfo> mesosTasks = new LinkedList<>();
+      for (LaunchableTask task : subTasks) {
+        Protos.TaskInfo pTask = task.constructMesosTaskInfo(heronConfig, heronRuntime);
+        mesosTasks.add(pTask);
+      }
+
+      LOG.info("Launching tasks from offer: " + offer + " with tasks: " + mesosTasks);
+
+      // Launch the tasks for the same offer
+      Protos.Status status =
+          driver.launchTasks(Arrays.asList(new Protos.OfferID[]{offer.getId()}),
+              mesosTasks);
+
+      if (status == Protos.Status.DRIVER_RUNNING) {
+        LOG.info(String.format("Tasks launched, status: '%s'", status));
+
+      } else {
+        LOG.info("Other status returned: " + status);
```

Comment:
Is any other status ok? Also in method `generateLaunchableTasks`, taskId is removed from `toScheduleTasks` if it has a matching offer. However if the status is not running, does that still mean that task is running. Otherwise the task should not be removed from `toScheduleTasks` till it is successfully scheduled.